env/posix/ocf_env.c

```diff
@@ -130,8 +131,57 @@ void env_stack_trace(void)
 }
 
 /* *** CRC *** */
-
 uint32_t env_crc32(uint32_t crc, uint8_t const *data, size_t len)
 {
 	return crc32(crc, data, len);
 }
+
+/* *** execution contexts *** */
+pthread_mutex_t exec_context_list_mutex = PTHREAD_MUTEX_INITIALIZER;
+pthread_mutex_t *exec_context_mutex = NULL;
+
+static void init_execution_context(void)
+{
+	unsigned count = env_get_execution_context_count();
+	unsigned i;
+
+	ENV_BUG_ON(count == -1);
+	exec_context_mutex = malloc(count * sizeof(exec_context_mutex[0]));
```

Comment:
Memory leak: this memory is never freed. Also mutexes created below are never deinitialized.
Maybe we should introduce `env_init()`/`env_deinit()` functions that would called before first and after last env_ call? Then we would call them during intialization/deinitialization of ocf_ctx (with some kind of refcount). Or even this refcount may be moved to env, and then we would have `env_get()` that initializes env if it's not initialized yet and increments refcount, and `env_put()` that decrements refcount and deinitializes env when refcount reaches zero.

Reply:
I think that moving env from stateless to stateful is a step in dangerous direction. What about ` __attribute__((destructor)) `? This way state specific to POSIX env is destroyed in POSIX specific way, at shared object unload or program end.

Reply:
Went with GCC constructor/destructor attributes to assure proper cleanup and initialization.